app/javascripts/main/packageManager.js

```diff
@@ -1,255 +1,267 @@
-
-    ipcMain.on('sync-components', (event, data) => {
-      this.syncComponents(data.componentsData);
-    });
-  }
-
-  pathsForComponent(component) {
-    const relativePath = `${ExtensionsFolderName}/` + component.content.package_info.identifier;
-    return {
-      downloadPath: appPath + `/${ExtensionsFolderName}/downloads/` + component.content.name + ".zip",
-      relativePath: relativePath,
-      absolutePath: appPath + "/" + relativePath
-    };
-  }
+const MappingFileLocation = `${appPath}/${ExtensionsFolderName}/mapping.json`;
+
+export function initializePackageManager(webContents) {
+  const syncTasks = [];
+  let isRunningTasks = false;
+
+  ipcMain.on(SYNC_COMPONENTS, async (_event, data) => {
+    console.log("Package manager: received event", SYNC_COMPONENTS);
+    syncTasks.push(data.componentsData);
+    if (isRunningTasks) return;
+    isRunningTasks = true;
+    await runTasks(webContents, syncTasks);
+    isRunningTasks = false;
+  });
+}
 
-  installComponent(component) {
-    const downloadUrl = component.content.package_info.download_url;
-    if (!downloadUrl) {
-      return;
-    }
+function findOppositeTask(tasks, { uuid, deleted }) {
+  return tasks.find(([t]) => t.uuid === uuid && t.deleted !== deleted);
+}
 
-    console.log("Installing component", component.content.name, downloadUrl);
-
-    const callback = (installedComponent, error) => {
-      this.window.webContents.send("install-component-complete", { component: installedComponent, error: error });
-    };
-
-    const paths = this.pathsForComponent(component);
-
-    fileUtils.downloadFile(downloadUrl, paths.downloadPath, (error) => {
-      if (!error) {
-        // Delete any existing content, especially in the case of performing an update
-        fileUtils.deleteAppRelativeDirectory(paths.relativePath);
-
-        // Extract contents
-        this.unzipFile(paths.downloadPath, paths.absolutePath, (err) => {
-          if (!err) {
-            this.unnestPackageContents(paths.absolutePath, () => {
-              // Find out main file
-              fileUtils.readJSONFile(paths.absolutePath + "/package.json", (response, error) => {
-                var main;
-                if (response) {
-                  if (response.sn) { main = response.sn.main; }
-                  if (response.version) { component.content.package_info.version = response.version; }
-                }
-                if (!main) { main = "index.html"; }
-
-                component.content.local_url = "sn://" + paths.relativePath + "/" + main;
-                callback(component);
-
-                // Update mapping file
-                this.updateMappingFile(component.uuid, paths.relativePath);
-              });
-            });
+async function runTasks(webContents, syncTasks) {
+  const tries = 3;
+  while (syncTasks.length > 0) {
+    try {
+      // Try to execute the task with up to three tries.
+      for (let triesAmount = 1; triesAmount <= 3; triesAmount++) {
```

Comment:
Should be using `tries` const here instead of `3`?

Reply:
This for loop syntax is fine, but in my experience for loops start at 0, with the middle constraint being a < operator rather than <=.

Reply:
I'd like to keep it that way because of the code in the `catch` block:
```
if (tries < maxTries) {
  continue;
} else {
  throw error;
}
```
If we start at zero we'd have to check for `tries+1<maxTries`, which I thought was awkward.